### DISCUSSION TITLE
XCframework compilation fails due to macros

### DISCUSSION THREAD

**Discussion by miguelkennedy** [pointfreeco/swift-composable-architecture#2639] [Q&A]
Hello everyone.

I use the ComposableArchitecture in my project as a xcframework and imported as a binary target. I've been using it like this for some time without any issues.  It allows faster compile times in my project.

I haven't updated the ComposableArchitecture dependency for quite some time and today I decided to update it so I could also play with the new macro features. I can't seem to compile it into an xcframework anymore, I get the error: 

`"Internal error. Please file a bug at https://github.com/apple/swift-package-manager/issues with this info. macro not supported"**`

@flockoffiles I know you also use The Composable Architecture in your project as a xcframework. Are you able to compile the latest version?

Does anyone want to chime in?

For reference I'm using the following command to try and compile it:

`swift create-xcframework ComposableArchitecture CasePaths CombineSchedulers ConcurrencyExtras CustomDump Dependencies DependenciesMacros IdentifiedCollections OrderedCollections SwiftUINavigationCore XCTestDynamicOverlay --clean --configuration release --debug-symbols --stack-evolution --platform ios --platform macos --platform tvos --platform watchos --output xcframeworks`

**Comment by flockoffiles** [comment]
Unfortunately I have not had a chance to try the new composable architecture version. I hope, @stephencelis  and @mbrandonw will get some time to react to this discussion at some point. (They have been very helpful in trying to resolve compilation issues in the past, but beware that it can also be some intrinsic Apple problem). I remember finding a problem in their generic frameworks, and Apple folks very kind to eventually implement some workaround to get it to compile. Not sure about this one though... (May take some time if it's on the Apple's side).

**Reply by stephencelis** [reply]
We don't have a ton of experience with this, but believe that Tuist has made recent changes to support macros. It might be worth checking with that project and also opening a discussion on the official Swift forums.

**Comment by darrarski** [comment]
I was able to make an XCFramework for Composable Architecture (and its dependencies) using [Scipio](https://github.com/giginet/Scipio). 

1. I've created a new swift package:

   ```sh
   mkdir MyDependencies
   cd MyDependencies
   swift package init
   ```

2. Then updated `Package.swift` to look like this:
   ```swift
   // swift-tools-version: 5.10
   import PackageDescription

   let package = Package(
     name: "MyDependencies",
     platforms: [
       .iOS(.v15),
     ],
     products: [],
     dependencies: [
       .package(url: "https://github.com/pointfreeco/swift-composable-architecture.git", from: "1.10.2"),
     ],
     targets: [
       .target(
         name: "MyDependencies",
         dependencies: [
           .product(name: "ComposableArchitecture", package: "swift-composable-architecture"),
         ]
       ),
     ]
   )
   ```
3. Finally, I run [Scipio](https://github.com/giginet/Scipio) from the package directory:
   ```sh
   scipio prepare . --support-simulators
   ```
4. XCFrameworks were generated successfully and stored in `XCFrameworks` directory.

While it generated XCFrameworks without any issues, and the frameworks seem to work fine when used in an Xcode project, the macros unfortunately do not work. When trying to use `@Reducer` I get the following error:

> External macro implementation type 'ComposableArchitectureMacros.ReducerMacro' could not be found for macro 'Reducer()'

I'm not sure how to fix the problem. I've opened [an issue](https://github.com/giginet/Scipio/issues/117) regarding it, and looking for support.

Besides this, I haven't found any problems so far. I'm able to use the Composable Architecture distributed in an XCFramework.

**Reply by darrarski** [reply]
I made some progress regarding the Composable Architecture XCFramrwork with Swift Macros. It's nowhere close to a proper solution, but I made it work with some manual hacking, so I guess it's worth sharing.

1. I built the Composable Architecture package.
2. I copied these binary files from the build / derived data directory into some directory, so I can find them later easily:
   - `CasePathsMacros`
   - `ComposableArchitectureMacros`
   - `DependenciesMacrosPlugin`
   - `PerceptionMacros`
3. I switched to my Xcode project (the one that uses Composable Architecture XCFramework generated in the way described in my previous comment). 
4. I've added the following script as a build phase to my target. It copies the binary files from the directory mentioned in point 2. into the project's build directory:
   ```sh
   cp "path/to/some/directory/*Macros" "$BUILD_DIR/Debug$EFFECTIVE_PLATFORM_NAME/"
   ```
6. I've added the following "Other Swift Flags" (in the build settings) to my target:
   > -load-plugin-executable
   > $BUILD_DIR/Debug$EFFECTIVE_PLATFORM_NAME/ComposableArchitectureMacros#ComposableArchitectureMacros
   > -load-plugin-executable
   > $BUILD_DIR/Debug$EFFECTIVE_PLATFORM_NAME/DependenciesMacrosPlugin#DependenciesMacrosPlugin
   > -load-plugin-executable
   > $BUILD_DIR/Debug$EFFECTIVE_PLATFORM_NAME/PerceptionMacros#PerceptionMacros
   > -load-plugin-executable
   > $BUILD_DIR/Debug$EFFECTIVE_PLATFORM_NAME/CasePathsMacros#CasePathsMacros

Now, I'm able to use Composable Architecture XCFramework and use Swift Macros like `@Reducer` when running my app in iOS Simulator.

The only issue I found so far, is that when I use the macros in my code, Xcode previews are no longer working - fail to build due to the following error:

> external macro implementation type 'ComposableArchitectureMacros.ReducerMacro' could not be found for macro 'Reducer()'

I think the above manual process could be automated. The broken previews need investigation, it may be a matter of telling Xcode where to find the macro binaries when building for previews.

**Reply by darrarski** [reply]
I wrote a simple script that builds TCA and generates XCFrameworks + Swift Macros binaries. It works (mostly) fine when added as a dependency to a swift package, but there are some rough edges. Strangely, unit tests are not compiling for macOS but run and pass fine for iOS. `Reducer._printChanges` does not print anything when used within Xcode previews. Macros on the other hand seem to be working without an issue. You can check it out here:

https://github.com/darrarski/tca-frameworks

**Reply by mbrandonw** [reply]
> Reducer._printChanges does not print anything when used within Xcode previews

Aren't XCFrameworks built in release mode? All of the logging in `_printChanges` is hidden behind an `#if DEBUG` check.

We could think about changing that behavior in 2.0, or finally un-underscore it make it the new behavior.

**Reply by darrarski** [reply]
Oh, you are right. I totally forgot about this behaviour. I makes sense, it worked fine before, when I built the frameworks with debug configuration.
